provider/azure/environ.go

```diff
@@ -1074,7 +1074,56 @@ func (env *azureEnviron) instances(
 
 // AdoptResources is part of the Environ interface.
 func (env *azureEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
-	return errors.NotImplementedf("AdoptResources")
+	groupClient := resources.GroupsClient{env.resources}
+	resourceClient := resources.Client{env.resources}
+	var failed []string
+
+	resources, err := groupClient.ListResources(env.resourceGroup, "", nil)
+	if err != nil {
+		return errors.Annotate(err, "listing resources")
+	}
+	for resources.Value != nil {
+		for _, resource := range *resources.Value {
+
+			providerNamespace, parentPath, err := splitResourceType(*resource.Type)
+			if err != nil {
+				failed = append(failed, *resource.Name)
+				logger.Errorf("error getting namespace and path for type %q: %v", *resource.Type, err)
+			}
+
+			(*resource.Tags)[tags.JujuController] = &controllerUUID
```

Comment:
Good point - I meant to do that check before updating but forgot.

I'm not sure what you mean about reentrancy - you mean, could this be called multiple times with the same controller uuid? Or different ones? I guess I can see it being called more than once with the same uuid, if there was something wrong with the migration master worker.

Do you mean, if it was called in parallel with two different uuids could we end up in an inconsistent state? If the order of the resources coming back was inconsistent between two different calls I think that would cause a problem. Hmm. Or if one of them was delayed for a long time and then continued after the other one had run.

I'm not sure how to handle that - I can think of a couple of options that still seem racy, unless there's a way to make the update conditional on the value of the tag being what we expect (on the azure end rather than ours).

I don't think the migration machinery would allow this to happen (concurrent calls with different uuids) - the second migration would be prevented until the first was finished. Do you think that's good enough or should we try to handle that at this level?

(Or have I completely misunderstood the question? :)